src/app/fdctl/run/run.c

```diff
@@ -417,83 +232,143 @@ main_pid_namespace( void * args ) {
+          exit_group( 0 );
+        }
+
+        /* Child process died, reap it to figure out exit code. */
+        int wstatus;
+        int exited_pid = wait4( -1, &wstatus, (int)__WALL | (int)WNOHANG, NULL );
+        if( FD_UNLIKELY( -1==exited_pid ) ) {
+          fd_log_private_fprintf_nolock_0( STDERR_FILENO, "pidns wait4() failed (%i-%s) %lu %hu", errno, fd_io_strerror( errno ), i, fds[i].revents );
+          exit_group( 1 );
+        } else if( FD_UNLIKELY( !exited_pid ) ) {
+          /* Spurious wakeup, no child actually dead yet. */
+          continue;
+        }
+
+        char * tile_name = child_names[ i-1 ];
+        ulong  tile_id = i-1 ? config->topo.tiles[ i-1 ].kind_id : 0;
+
+        if( FD_UNLIKELY( !WIFEXITED( wstatus ) ) ) {
+          fd_log_private_fprintf_nolock_0( STDERR_FILENO, "tile %s:%lu exited with signal %d (%s)\n", tile_name, tile_id, WTERMSIG( wstatus ), fd_io_strsignal( WTERMSIG( wstatus ) ) );
+          exit_group( WTERMSIG( wstatus ) ? WTERMSIG( wstatus ) : 1 );
+        }
+        fd_log_private_fprintf_nolock_0( STDERR_FILENO, "tile %s:%lu exited with code %d\n", tile_name, tile_id, WEXITSTATUS( wstatus ) );
+        exit_group( WEXITSTATUS( wstatus ) ? WEXITSTATUS( wstatus ) : 1 );
+      }
+    }
   }
-  fd_log_private_fprintf_nolock_0( STDERR_FILENO, "tile %lu (%s) exited with code %d\n", tile_idx, name, WEXITSTATUS( wstatus ) );
-  exit_group( WEXITSTATUS( wstatus ) ? WEXITSTATUS( wstatus ) : 1 );
   return 0;
 }
 
-static pid_t pid_namespace;
-extern char fd_log_private_path[ 1024 ]; /* empty string on start */
+/* The boot sequence is a little bit involved...
 
-static void
-parent_signal( int sig ) {
-  (void)sig;
-  if( FD_LIKELY( pid_namespace ) ) kill( pid_namespace, SIGKILL );
-  fd_log_private_fprintf_nolock_0( STDERR_FILENO, "Log at \"%s\"\n", fd_log_private_path );
-  exit_group( 0 );
-}
+   A process tree is created that looks like,
 
-static void
-install_parent_signals( void ) {
-  struct sigaction sa = {
-    .sa_handler = parent_signal,
-    .sa_flags   = 0,
-  };
-  if( FD_UNLIKELY( sigaction( SIGTERM, &sa, NULL ) ) )
-    FD_LOG_ERR(( "sigaction(SIGTERM) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
-  if( FD_UNLIKELY( sigaction( SIGINT, &sa, NULL ) ) )
-    FD_LOG_ERR(( "sigaction(SIGINT) failed (%i-%s)", errno, fd_io_strerror( errno ) ));
-}
+   + main
+   +-- pidns
+       +-- solana-labs
+       +-- tile 0
+       +-- tile 1
+       ...
+
+   What we want is that if any process in the tree dies, all other
+   processes will also die.  This is done as follows,
+
+    (a) pidns is the init process of a PID namespace, so if it dies the
+        kernel will terminate the child processes.
+
+    (b) main is the parent of pidns, so it can issue a waitpid() on the
+        child PID, and when it completes terminate itself.
+
+    (c) pidns is the parent of solana-labs and the tiles, so it could
+        issue a waitpid() of -1 to wait for any of them to terminate,
```

Comment:
If it can do this, why do you need the child->pidns pipe?

Reply:
It makes the logic more difficult, you end up with something like

```while(1) { if wait(-1, NONBLOCK) { ..child died } else if (poll(parent_pipe_fd, NONBLOCK)) ... }```

so you have to spin a core and so on. Better just issue one syscall and let the kernel deal with it.